File: src/Colisores/Fisica.c
// Fisica.c
// TupiEngine - Sistema de Física

#include "Fisica.h"
#include <math.h>

// ============================================================
// MOVIMENTO
// ============================================================

void tupi_fisica_atualizar(TupiCorpo* c, float dt, float gravidade) {
    if (c->massa == 0.0f) return;

    c->aceleracaoY += gravidade;

    c->velX += c->aceleracaoX * dt;
    c->velY += c->aceleracaoY * dt;

    c->x += c->velX * dt;
    c->y += c->velY * dt;

    c->aceleracaoX = 0.0f;
    c->aceleracaoY = 0.0f;
}

void tupi_fisica_impulso(TupiCorpo* c, float fx, float fy) {
    if (c->massa == 0.0f) return;
    c->velX += fx / c->massa;
    c->velY += fy / c->massa;
}

// ============================================================
// HELPERS DE FORMA
// ============================================================

TupiRetCol tupi_corpo_ret(TupiCorpo* c, float largura, float altura) {
    return tupi_ret(c->x, c->y, largura, altura);
}

TupiCircCol tupi_corpo_cir(TupiCorpo* c, float raio) {
    return tupi_cir(c->x, c->y, raio);
}

// ============================================================
// RESOLUÇÃO DE COLISÃO
// ============================================================

void tupi_resolver_colisao(TupiCorpo* a, TupiCorpo* b, TupiColisao info) {
    if (!info.colidindo) return;

    // Separa proporcionalmente às massas
    float massaTotal = a->massa + b->massa;
    float ratioA = (b->massa > 0.0f) ? b->massa / massaTotal : 1.0f;
    float ratioB = (a->massa > 0.0f) ? a->massa / massaTotal : 1.0f;

    if (a->massa > 0.0f) {
        a->x += info.dx * ratioA;
        a->y += info.dy * ratioA;
    }
    if (b->massa > 0.0f) {
        b->x -= info.dx * ratioB;
        b->y -= info.dy * ratioB;
    }

    // Resposta de velocidade com elasticidade média
    float e = (a->elasticidade + b->elasticidade) * 0.5f;

    if (info.dy != 0.0f) {
        float velRelY = a->velY - b->velY;
        float impulso = -(1.0f + e) * velRelY / (1.0f/a->massa + 1.0f/b->massa);
        if (a->massa > 0.0f) a->velY += impulso / a->massa;
        if (b->massa > 0.0f) b->velY -= impulso / b->massa;
    }
    if (info.dx != 0.0f) {
        float velRelX = a->velX - b->velX;
        float impulso = -(1.0f + e) * velRelX / (1.0f/a->massa + 1.0f/b->massa);
        if (a->massa > 0.0f) a->velX += impulso / a->massa;
        if (b->massa > 0.0f) b->velX -= impulso / b->massa;
    }
}

void tupi_resolver_estatico(TupiCorpo* a, TupiColisao info) {
    if (!info.colidindo || a->massa == 0.0f) return;

    a->x += info.dx;
    a->y += info.dy;

    if (info.dy != 0.0f) a->velY = -a->velY * a->elasticidade;
    if (info.dx != 0.0f) a->velX = -a->velX * a->elasticidade;
}

// ============================================================
// UTILITÁRIOS
// ============================================================

void tupi_aplicar_atrito(TupiCorpo* c, float dt) {
    if (c->massa == 0.0f) return;
    float fator = 1.0f - c->atrito * dt;
    if (fator < 0.0f) fator = 0.0f;
    c->velX *= fator;
    c->velY *= fator;
}

void tupi_limitar_velocidade(TupiCorpo* c, float maxVel) {
    float mag2 = c->velX * c->velX + c->velY * c->velY;
    if (mag2 > maxVel * maxVel) {
        float escala = maxVel / sqrtf(mag2);
        c->velX *= escala;
        c->velY *= escala;
    }
}

File: src/Engine/fisica.lua
-- fisica.lua — Física 2D (corpos dinâmicos, impulsos, resolução de colisão).

local ffi = require("ffi")
local C   = require("src.Engine.engineffi")

local M = {}

-- ============================================================
-- FÍSICA 2D
-- ============================================================

--- Cria um corpo físico dinâmico (afetado pela gravidade).
-- @param x, y          posição inicial
-- @param massa         valores típicos: 1.0–100.0
-- @param elasticidade  0.0 (para) a 1.0 (ricochete perfeito)
-- @param atrito        0.0 (desliza) a 1.0 (para rápido)
-- @return              cdata TupiCorpo[1]
function M.corpo(x, y, massa, elasticidade, atrito)
    local c = ffi.new("TupiCorpo[1]")
    c[0].x            = x            or 0
    c[0].y            = y            or 0
    c[0].velX         = 0
    c[0].velY         = 0
    c[0].aceleracaoX  = 0
    c[0].aceleracaoY  = 0
    c[0].massa        = massa        or 1.0
    c[0].elasticidade = elasticidade or 0.3
    c[0].atrito       = atrito       or 0.1
    return c
end

--- Cria um corpo estático (massa=0, nunca se move).
function M.corpoEstatico(x, y)
    local c = ffi.new("TupiCorpo[1]")
    c[0].x            = x or 0
    c[0].y            = y or 0
    c[0].velX         = 0
    c[0].velY         = 0
    c[0].aceleracaoX  = 0
    c[0].aceleracaoY  = 0
    c[0].massa        = 0.0
    c[0].elasticidade = 0.0
    c[0].atrito       = 1.0
    return c
end

--- Avança a simulação do corpo por um passo de tempo.
-- @param gravidade  aceleração em px/s² (padrão 500.0)
function M.atualizar(corpo, gravidade)
    -- Depende de Tupi.dt() — injetado via M._getDt ao montar o TupiEngine
    local dt = M._getDt and M._getDt() or 0
    C.tupi_fisica_atualizar(corpo, dt, gravidade or 500.0)
end

--- Aplica impulso instantâneo (pulo, explosão, knockback).
function M.impulso(corpo, fx, fy)
    C.tupi_fisica_impulso(corpo, fx or 0, fy or 0)
end

--- Retorna o retângulo de colisão centrado no corpo.
function M.retCol(corpo, largura, altura)
    local r = C.tupi_corpo_ret(corpo, largura or 0, altura or 0)
    return { x = tonumber(r.x), y = tonumber(r.y),
             largura = tonumber(r.largura), altura = tonumber(r.altura) }
end

--- Retorna o círculo de colisão centrado no corpo.
function M.cirCol(corpo, raio)
    local c = C.tupi_corpo_cir(corpo, raio or 0)
    return { x = tonumber(c.x), y = tonumber(c.y), raio = tonumber(c.raio) }
end

--- Resolve colisão entre dois corpos dinâmicos.
function M.resolverColisao(a, b, info)
    local c = ffi.new("TupiColisao",
        info.colidindo and 1 or 0,
        info.dx or 0,
        info.dy or 0)
    C.tupi_resolver_colisao(a, b, c)
end

--- Resolve colisão entre um corpo dinâmico e um obstáculo estático.
function M.resolverEstatico(corpo, info)
    local c = ffi.new("TupiColisao",
        info.colidindo and 1 or 0,
        info.dx or 0,
        info.dy or 0)
    C.tupi_resolver_estatico(corpo, c)
end

function M.atrito(corpo)
    local dt = M._getDt and M._getDt() or 0
    C.tupi_aplicar_atrito(corpo, dt)
end
function M.limitarVel(corpo, maxVel) C.tupi_limitar_velocidade(corpo, maxVel or 800.0) end
function M.pos(corpo)  return tonumber(corpo[0].x),    tonumber(corpo[0].y)    end
function M.vel(corpo)  return tonumber(corpo[0].velX), tonumber(corpo[0].velY) end

function M.setPosicao(corpo, x, y)
    corpo[0].x = x or corpo[0].x
    corpo[0].y = y or corpo[0].y
end

function M.setVel(corpo, vx, vy)
    corpo[0].velX = vx or 0
    corpo[0].velY = vy or 0
end

--- Sincroniza a posição de um wrapper com seu TupiCorpo.
function M.sincronizar(wrapper, corpo)
    wrapper.obj[0].x = corpo[0].x
    wrapper.obj[0].y = corpo[0].y
end

return M
